File: src/com/leozhi/topic392/Solution04.java
package com.leozhi.topic392;

/**
 * @author leozhi
 * 双指针
 * 通过
 * 2ms
 */
public class Solution04 {
    public boolean isSubsequence(String s, String t) {
        if (s.length() == 0) {
            return true;
        } else if(t.length() == 0) {
            return false;
        }
        int i = 0, j = 0;
        while (i < s.length()) {
            j = i;
            while (j < t.length()) {
                if (s.charAt(i) == t.charAt(j)) {
                    i++;
                    break;
                } else {
                    if (++j == t.length()) {
                        return false;
                    }
                }
            }
        }
        return true;
    }
}

File: src/com/leozhi/topic861/Solution04.kt
package com.leozhi.topic861

import kotlin.math.max

/**
 * @author leozhi
 * 贪心算法
 * 通过
 * 208ms
 */
class Solution04 {
    fun matrixScore(A: Array<IntArray>): Int {
        val m = A.size
        val n = A[0].size
        var res = m * (1.shl(n - 1))
        for (i in 1 until n) {
            var nOnes = 0
            for (ints in A) {
                if (ints[0] == 1) {
                    nOnes += ints[i]
                } else {
                    nOnes += (1 - ints[i])
                }
            }
            val max = max(nOnes, m - nOnes)
            res += max * (1.shl(n - i - 1))
        }
        return res
    }
}